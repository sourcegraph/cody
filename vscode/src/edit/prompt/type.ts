import type { PromptString, Rule } from '@sourcegraph/cody-shared'
import type * as vscode from 'vscode'

export interface LLMPrompt {
    system?: PromptString
    instruction: PromptString
}

export interface LLMInteraction {
    prompt: LLMPrompt
    stopSequences?: string[]
    responseTopic?: PromptString
    assistantText?: PromptString
    assistantPrefix?: PromptString
}

export interface GetLLMInteractionOptions {
    instruction: PromptString
    precedingText: PromptString
    selectedText: PromptString
    followingText: PromptString
    uri: vscode.Uri
    document: vscode.TextDocument
    rules?: Rule[] | null
}

type LLMInteractionBuilder = (options: GetLLMInteractionOptions) => LLMInteraction

export interface EditLLMInteraction {
    getEdit: LLMInteractionBuilder
    getDoc: LLMInteractionBuilder
    getFix: LLMInteractionBuilder
    getAdd: LLMInteractionBuilder
    getTest: LLMInteractionBuilder
}
