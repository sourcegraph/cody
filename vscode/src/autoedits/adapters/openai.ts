import { autoeditsLogger } from '../logger'
import type { AutoeditModelOptions, AutoeditsModelAdapter } from './base'
import { getModelResponse, getOpenaiCompatibleChatPrompt } from './utils'

export class OpenAIAdapter implements AutoeditsModelAdapter {
    async getModelResponse(option: AutoeditModelOptions): Promise<string> {
        try {
            const response = await getModelResponse(
                option.url,
                JSON.stringify({
                    model: option.model,
                    messages: getOpenaiCompatibleChatPrompt({
                        systemMessage: option.prompt.systemMessage,
                        userMessage: option.prompt.userMessage,
                    }),
                    temperature: 0.5,
                    max_tokens: 256,
                    response_format: {
                        type: 'text',
                    },
                }),
                option.apiKey
            )
            return response.choices[0].message.content
        } catch (error) {
            autoeditsLogger.logDebug('AutoEdits', 'Error calling OpenAI API:', error)
            throw error
        }
    }
}
