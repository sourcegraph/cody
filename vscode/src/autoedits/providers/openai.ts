import type { AutoEditsTokenLimit } from '@sourcegraph/cody-shared'
import type * as vscode from 'vscode'
import type {
    AutocompleteContextSnippet,
    DocumentContext,
} from '../../../../lib/shared/src/completions/types'
import { autoeditsLogger } from '../logger'
import type {
    ChatPrompt,
    PromptProvider,
    PromptProviderResponse,
    PromptResponseData,
} from '../prompt-provider'
import { getModelResponse } from '../prompt-provider'
import { type CodeToReplaceData, SYSTEM_PROMPT, getBaseUserPrompt } from '../prompt-utils'

export class OpenAIPromptProvider implements PromptProvider {
    getPrompt(
        docContext: DocumentContext,
        document: vscode.TextDocument,
        position: vscode.Position,
        context: AutocompleteContextSnippet[],
        tokenBudget: AutoEditsTokenLimit
    ): PromptResponseData {
        const { codeToReplace, promptResponse: userPrompt } = getBaseUserPrompt(
            docContext,
            document,
            position,
            context,
            tokenBudget
        )
        const prompt: ChatPrompt = [
            {
                role: 'system',
                content: SYSTEM_PROMPT,
            },
            {
                role: 'user',
                content: userPrompt,
            },
        ]
        return {
            codeToReplace: codeToReplace,
            promptResponse: prompt,
        }
    }

    postProcessResponse(codeToReplace: CodeToReplaceData, response: string): string {
        return response
    }

    async getModelResponse(
        model: string,
        apiKey: string,
        prompt: PromptProviderResponse
    ): Promise<string> {
        try {
            const response = await getModelResponse(
                'https://api.openai.com/v1/chat/completions',
                JSON.stringify({
                    model: model,
                    messages: prompt,
                    temperature: 0.5,
                    max_tokens: 256,
                    response_format: {
                        type: 'text',
                    },
                }),
                apiKey
            )
            return response.choices[0].message.content
        } catch (error) {
            autoeditsLogger.logDebug('AutoEdits', 'Error calling OpenAI API:', error)
            throw error
        }
    }
}
