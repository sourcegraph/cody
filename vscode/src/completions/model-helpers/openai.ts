import { type DocumentContext, type PromptString, ps } from '@sourcegraph/cody-shared'

import {
    CLOSING_CODE_TAG,
    OPENING_CODE_TAG,
    extractFromCodeBlock,
    fixBadCompletionStart,
    getHeadAndTail,
    trimLeadingWhitespaceUntilNewline,
} from '../text-processing'
import {
    DefaultModel,
    type FormatPromptParams,
    type GetDefaultIntroSnippetsParams,
    type GetOllamaPromptParams,
} from './default'

export class OpenAIV1 extends DefaultModel {
    public stopSequences = [CLOSING_CODE_TAG.toString()]

    private instructions =
        ps`You are a code completion AI designed to take the surrounding code and shared context into account in order to predict and suggest high-quality code to complete the code enclosed in ${OPENING_CODE_TAG} tags.  You only respond with code that works and fits seamlessly with surrounding code. Do not include anything else beyond the code.`

    getOllamaPrompt(promptContext: GetOllamaPromptParams): PromptString {
        throw new Error('OpenAI is not supported by the Ollama provider yet!')
    }

    protected getDefaultIntroSnippets(params: GetDefaultIntroSnippetsParams): PromptString[] {
        return [this.instructions]
    }

    formatPrompt(params: FormatPromptParams): PromptString {
        const { intro, prefix, suffix, fileName } = params

        const { head, tail } = getHeadAndTail(prefix)
        const infillBlock = tail.trimmed.toString().endsWith('{\n')
            ? tail.trimmed.trimEnd()
            : tail.trimmed

        const infillPrefix = head.raw || ''

        return ps`Below is the code from file path ${fileName}. Review the code outside the XML tags to detect the functionality, formats, style, patterns, and logics in use. Then, use what you detect and reuse methods/libraries to complete and enclose completed code only inside XML tags precisely without duplicating existing implementations. Here is the code:\n\`\`\`\n${intro}${infillPrefix}${OPENING_CODE_TAG}${CLOSING_CODE_TAG}${suffix}\n\`\`\`

${OPENING_CODE_TAG}${infillBlock}`
    }

    public postProcess(content: string, docContext: DocumentContext): string {
        let completion = extractFromCodeBlock(content)

        const trimmedPrefixContainNewline = docContext.prefix
            .slice(docContext.prefix.trimEnd().length)
            .includes('\n')
        if (trimmedPrefixContainNewline) {
            // The prefix already contains a `\n` that LLM was not aware of, so we remove any
            // leading `\n` followed by whitespace that might be add.
            completion = completion.replace(/^\s*\n\s*/, '')
        } else {
            completion = trimLeadingWhitespaceUntilNewline(completion)
        }

        // Remove bad symbols from the start of the completion string.
        completion = fixBadCompletionStart(completion)

        return completion
    }
}
