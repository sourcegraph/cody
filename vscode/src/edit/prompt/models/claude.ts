import { ps } from '@sourcegraph/cody-shared'
import { PROMPT_TOPICS } from '../constants'
import type { EditLLMInteraction } from '../type'
import { buildGenericPrompt } from './generic'

const RESPONSE_PREFIX = ps`<${PROMPT_TOPICS.OUTPUT}>`
const TEST_FILE_PREFIX = ps`<${PROMPT_TOPICS.FILENAME}>`
const SHARED_PARAMETERS = {
    responseTopic: PROMPT_TOPICS.OUTPUT,
    stopSequences: [`</${PROMPT_TOPICS.OUTPUT}>`],
    assistantText: RESPONSE_PREFIX,
    assistantPrefix: RESPONSE_PREFIX,
}

export const claude: EditLLMInteraction = {
    getEdit(options) {
        return {
            ...SHARED_PARAMETERS,
            prompt: buildGenericPrompt('edit', options),
        }
    },
    getDoc(options) {
        const docStopSequences = [...SHARED_PARAMETERS.stopSequences]
        const firstLine = options.selectedText.toString().split('\n')[0]
        // when claude receives a def function line, it will return an empty response
        // if we add that first line to the stop sequences
        if (firstLine.trim().length > 0 && options.document.languageId !== 'python') {
            docStopSequences.push(firstLine)
        }

        return {
            ...SHARED_PARAMETERS,
            stopSequences: docStopSequences,
            prompt: buildGenericPrompt('doc', options),
        }
    },
    getFix(options) {
        return {
            ...SHARED_PARAMETERS,
            prompt: buildGenericPrompt('fix', options),
        }
    },
    getAdd(options) {
        let assistantPreamble = ps``
        if (options.precedingText.length > 0) {
            assistantPreamble = ps`<${PROMPT_TOPICS.PRECEDING}>${options.precedingText}</${PROMPT_TOPICS.PRECEDING}>`
        }
        return {
            ...SHARED_PARAMETERS,
            assistantText: ps`${assistantPreamble}${RESPONSE_PREFIX}`,
            prompt: buildGenericPrompt('add', options),
        }
    },
    getTest(options) {
        return {
            ...SHARED_PARAMETERS,
            assistantText: ps`${RESPONSE_PREFIX}${TEST_FILE_PREFIX}`,
            assistantPrefix: ps`${RESPONSE_PREFIX}${TEST_FILE_PREFIX}`,
            prompt: buildGenericPrompt('test', options),
        }
    },
}
