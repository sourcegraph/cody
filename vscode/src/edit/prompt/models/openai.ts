import { ps } from '@sourcegraph/cody-shared'
import { PROMPT_TOPICS } from '../constants'
import type { EditLLMInteraction } from '../type'
import { buildGenericPrompt, getGenericPrefixes } from './generic'

const RESPONSE_PREFIX = ps`<${PROMPT_TOPICS.OUTPUT}>\n`
const SHARED_PARAMETERS = {
    responseTopic: PROMPT_TOPICS.OUTPUT,
    stopSequences: [`</${PROMPT_TOPICS.OUTPUT}>`],
}

const MODEL_PREFIX = {
    assistantText: RESPONSE_PREFIX,
    assistantPrefix: RESPONSE_PREFIX,
} as const

export const openai: EditLLMInteraction = {
    getEdit(options) {
        return {
            ...SHARED_PARAMETERS,
            ...getGenericPrefixes(MODEL_PREFIX, options.isReasoningModel),
            prompt: buildGenericPrompt('edit', options),
        }
    },
    getDoc(options) {
        return {
            ...SHARED_PARAMETERS,
            ...getGenericPrefixes(MODEL_PREFIX, options.isReasoningModel),
            prompt: buildGenericPrompt('doc', options),
        }
    },
    getFix(options) {
        return {
            ...SHARED_PARAMETERS,
            ...getGenericPrefixes(MODEL_PREFIX, options.isReasoningModel),
            prompt: buildGenericPrompt('fix', options),
        }
    },
    getAdd(options) {
        let assistantPreamble = ps``
        if (options.precedingText.length > 0) {
            assistantPreamble = ps`<${PROMPT_TOPICS.PRECEDING}>${options.precedingText}</${PROMPT_TOPICS.PRECEDING}>`
        }
        return {
            ...SHARED_PARAMETERS,
            ...getGenericPrefixes(
                { ...MODEL_PREFIX, assistantText: ps`${assistantPreamble}${RESPONSE_PREFIX}` },
                options.isReasoningModel
            ),
            prompt: buildGenericPrompt('add', options),
        }
    },
    getTest(options) {
        return {
            ...SHARED_PARAMETERS,
            ...getGenericPrefixes(MODEL_PREFIX, options.isReasoningModel),
            prompt: buildGenericPrompt('test', options),
        }
    },
}
