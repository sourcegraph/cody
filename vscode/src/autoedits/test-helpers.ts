import { vi } from 'vitest'
import type * as vscode from 'vscode'

import type { ChatClient } from '@sourcegraph/cody-shared'

import { documentAndPosition } from '../completions/test-helpers'
import { defaultVSCodeExtensionClient } from '../extension-client'
import { FixupController } from '../non-stop/FixupController'
import { WorkspaceEdit, vsCodeMocks } from '../testutils/mocks'

import type { CodyStatusBar } from '../services/StatusBar'
import * as adapters from './adapters/utils'
import { autoeditTriggerKind } from './analytics-logger'
import {
    AUTOEDIT_TOTAL_DEBOUNCE_INTERVAL,
    AutoeditsProvider,
    type AutoeditsResult,
} from './autoedits-provider'

/**
 * A helper to be used for the autoedits integration tests.
 *
 * Creates a mock environment and returns the autoedits result object.
 * Simulates VSCode's text editor, document context, and model responses with
 * configurable prediction text and completion context.
 */
export async function autoeditResultFor(
    textWithCursor: string,
    {
        inlineCompletionContext = {
            triggerKind: autoeditTriggerKind.automatic,
            selectedCompletionInfo: undefined,
        },
        prediction,
        token,
        provider: existingProvider,
        getModelResponse,
        isAutomaticTimersAdvancementDisabled = false,
    }: {
        prediction: string
        /** provide to reuse an existing provider instance */
        provider?: AutoeditsProvider
        inlineCompletionContext?: vscode.InlineCompletionContext
        token?: vscode.CancellationToken
        getModelResponse?: (
            url: string,
            body: string,
            apiKey: string,
            customHeaders?: Record<string, string>
        ) => Promise<unknown>
        isAutomaticTimersAdvancementDisabled?: boolean
    }
): Promise<{
    result: AutoeditsResult | null
    promiseResult: Promise<AutoeditsResult | null>
    document: vscode.TextDocument
    position: vscode.Position
    provider: AutoeditsProvider
    editBuilder: WorkspaceEdit
}> {
    const getModelResponseMock = async (...args: unknown[]) => {
        // Simulate response latency.
        vi.advanceTimersByTime(100)

        return {
            choices: [
                {
                    text: prediction,
                },
            ],
        }
    }

    // TODO: add a callback to verify `getModelResponse` arguments.
    vi.spyOn(adapters, 'getModelResponse').mockImplementation(getModelResponse || getModelResponseMock)

    const editBuilder = new WorkspaceEdit()
    const { document, position } = documentAndPosition(textWithCursor)

    vi.spyOn(vsCodeMocks.window, 'activeTextEditor', 'get').mockReturnValue({
        document,
        selection: {
            active: position,
        },
        edit: (callback: any) => callback(editBuilder),
        setDecorations: () => {},
    } as any)

    const chatClient = null as unknown as ChatClient
    const extensionClient = defaultVSCodeExtensionClient()
    const fixupController = new FixupController(extensionClient)
    const mockStatusBar = {
        addLoader: vi.fn(),
        init: vi.fn(),
    } as any as CodyStatusBar
    const provider =
        existingProvider ??
        new AutoeditsProvider(chatClient, fixupController, mockStatusBar, { shouldRenderImage: false })

    let result: AutoeditsResult | null = null

    const promiseResult = provider
        .provideInlineCompletionItems(document, position, inlineCompletionContext, token)
        .then(res => {
            result = res
            return result
        })

    if (!isAutomaticTimersAdvancementDisabled) {
        // Advance time by the default debounce interval.
        await vi.advanceTimersByTimeAsync(AUTOEDIT_TOTAL_DEBOUNCE_INTERVAL)
    }

    return { result, promiseResult, document, position, provider, editBuilder }
}
