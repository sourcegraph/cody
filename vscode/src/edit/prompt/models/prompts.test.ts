import { describe, expect, it } from 'vitest'

import { isWindows, ps, testFileUri } from '@sourcegraph/cody-shared'

import { document } from '../../../completions/test-helpers'
import type { GetLLMInteractionOptions } from '../type'
import { claude } from './claude'
import { openai } from './openai'

describe('Edit Prompts', () => {
    const followingText = ps`const text = 'Hello, world!'\n`
    const selectedText = ps`return text`
    const precedingText = ps`\n}`
    const uri = testFileUri('src/file/index.ts')
    const fixupTask: GetLLMInteractionOptions = {
        uri,
        followingText,
        selectedText,
        precedingText,
        instruction: ps`Console log text`,
        document: document(
            followingText.toString() + selectedText.toString() + precedingText.toString(),
            'typescript',
            uri.toString()
        ),
    }

    function normalize(text: string): string {
        return isWindows() ? text.replaceAll('src\\file\\index.ts', 'src/file/index.ts') : text
    }

    it.each([
        { name: 'claude', fn: claude },
        { name: 'openai', fn: openai },
    ])('$name builds prompts correctly', ({ fn }) => {
        const { prompt: editPrompt } = fn.getEdit(fixupTask)
        expect(normalize(editPrompt.system?.toString() || '')).toMatchSnapshot('edit.system')
        expect(normalize(editPrompt.instruction?.toString())).toMatchSnapshot('edit.instruction')

        // With rules provided.
        const { prompt: editPromptWithRules } = fn.getEdit({
            ...fixupTask,
            rules: [{ uri: 'file:///a.rule.md', display_name: 'a', instruction: 'My instruction' }],
        })
        expect(normalize(editPromptWithRules.system?.toString() || '')).toMatchSnapshot(
            'edit-with-rules.system'
        )
        expect(normalize(editPromptWithRules.instruction?.toString())).toMatchSnapshot(
            'edit-with-rules.instruction'
        )

        const { prompt: docPrompt } = fn.getDoc(fixupTask)
        expect(normalize(docPrompt.system?.toString() || '')).toMatchSnapshot('doc.system')
        expect(normalize(docPrompt.instruction?.toString())).toMatchSnapshot('doc.instruction')

        const { prompt: addPrompt } = fn.getAdd(fixupTask)
        expect(normalize(addPrompt.system?.toString() || '')).toMatchSnapshot('add.system')
        expect(normalize(addPrompt.instruction?.toString())).toMatchSnapshot('add.instruction')

        const { prompt: fixPrompt } = fn.getFix(fixupTask)
        expect(normalize(fixPrompt.system?.toString() || '')).toMatchSnapshot('fix.system')
        expect(normalize(fixPrompt.instruction?.toString())).toMatchSnapshot('fix.instruction')

        const { prompt: testPrompt } = fn.getTest(fixupTask)
        expect(normalize(testPrompt.system?.toString() || '')).toMatchSnapshot('test.system')
        expect(normalize(testPrompt.instruction?.toString())).toMatchSnapshot('test.instruction')
    })
})
