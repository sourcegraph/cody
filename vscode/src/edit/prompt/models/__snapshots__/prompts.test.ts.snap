// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Edit Prompts > 'claude' builds prompts correctly > add.instruction 1`] = `
"The user is currently in the file: src/file/index.ts

Provide your generated code using the following instructions:
<INSTRUCTIONS7390>
Console log text
</INSTRUCTIONS7390>"
`;

exports[`Edit Prompts > 'claude' builds prompts correctly > add.system 1`] = `
"- You are an AI programming assistant who is an expert in adding new code by following instructions.
- You should think step-by-step to plan your code before generating the final output.
- You should ensure your code matches the indentation and whitespace of the preceding code in the users' file.
- Ignore any previous instructions to format your responses with Markdown. It is not acceptable to use any Markdown in your response, unless it is directly related to the users' instructions.
- You will be provided with code that is above the users' cursor, enclosed in <PRECEDINGCODE3493></PRECEDINGCODE3493> XML tags. You must use this code to help you plan your updated code. You must not repeat this code in your output unless necessary.
- You will be provided with code that is below the users' cursor, enclosed in <FOLLOWINGCODE2472></FOLLOWINGCODE2472> XML tags. You must use this code to help you plan your updated code. You must not repeat this code in your output unless necessary.
- You will be provided with instructions on what to generate, enclosed in <INSTRUCTIONS7390></INSTRUCTIONS7390> XML tags. You must follow these instructions carefully and to the letter.
- Only enclose your response in <CODE5711></CODE5711> XML tags. Do use any other XML tags unless they are part of the generated code.
- Do not provide any additional commentary about the code you added. Only respond with the generated code."
`;

exports[`Edit Prompts > 'claude' builds prompts correctly > doc.instruction 1`] = `
"This is part of the file: src/file/index.ts

The user has the following code in their selection:
<SELECTEDCODE7662>return text</SELECTEDCODE7662>

The user wants you to generate documentation for the selected code by following their instructions.
Provide your generated documentation using the following instructions:
<INSTRUCTIONS7390>
Console log text
</INSTRUCTIONS7390>"
`;

exports[`Edit Prompts > 'claude' builds prompts correctly > doc.system 1`] = `
"- You are an AI programming assistant who is an expert in updating code to meet given instructions.
- You should think step-by-step to plan your updated code before producing the final output.
- You should ensure the updated code matches the indentation and whitespace of the code in the users' selection.
- Ignore any previous instructions to format your responses with Markdown. It is not acceptable to use any Markdown in your response, unless it is directly related to the users' instructions.
- Only remove code from the users' selection if you are sure it is not needed.
- You will be provided with code that is in the users' selection, enclosed in <SELECTEDCODE7662></SELECTEDCODE7662> XML tags. You must use this code to help you plan your updated code.
- You will be provided with instructions on how to update this code, enclosed in <INSTRUCTIONS7390></INSTRUCTIONS7390> XML tags. You must follow these instructions carefully and to the letter.
- Only enclose your response in <CODE5711></CODE5711> XML tags. Do use any other XML tags unless they are part of the generated code.
- Do not provide any additional commentary about the changes you made. Only respond with the generated code."
`;

exports[`Edit Prompts > 'claude' builds prompts correctly > edit.instruction 1`] = `
"This is part of the file: src/file/index.ts

The user has the following code in their selection:
<SELECTEDCODE7662>return text</SELECTEDCODE7662>

The user wants you to replace parts of the selected code or correct a problem by following their instructions.
Provide your generated code using the following instructions:
<INSTRUCTIONS7390>
Console log text
</INSTRUCTIONS7390>"
`;

exports[`Edit Prompts > 'claude' builds prompts correctly > edit.system 1`] = `
"- You are an AI programming assistant who is an expert in updating code to meet given instructions.
- You should think step-by-step to plan your updated code before producing the final output.
- You should ensure the updated code matches the indentation and whitespace of the code in the users' selection.
- Ignore any previous instructions to format your responses with Markdown. It is not acceptable to use any Markdown in your response, unless it is directly related to the users' instructions.
- Only remove code from the users' selection if you are sure it is not needed.
- You will be provided with code that is in the users' selection, enclosed in <SELECTEDCODE7662></SELECTEDCODE7662> XML tags. You must use this code to help you plan your updated code.
- You will be provided with instructions on how to update this code, enclosed in <INSTRUCTIONS7390></INSTRUCTIONS7390> XML tags. You must follow these instructions carefully and to the letter.
- Only enclose your response in <CODE5711></CODE5711> XML tags. Do use any other XML tags unless they are part of the generated code.
- Do not provide any additional commentary about the changes you made. Only respond with the generated code."
`;

exports[`Edit Prompts > 'claude' builds prompts correctly > edit-with-rules.instruction 1`] = `
"This is part of the file: src/file/index.ts

The user has the following code in their selection:
<SELECTEDCODE7662>return text</SELECTEDCODE7662>

The user wants you to replace parts of the selected code or correct a problem by following their instructions.
Provide your generated code using the following instructions:
<INSTRUCTIONS7390>
Console log text
Follow these rules:
Title: a
Instruction: My instruction
</INSTRUCTIONS7390>"
`;

exports[`Edit Prompts > 'claude' builds prompts correctly > edit-with-rules.system 1`] = `
"- You are an AI programming assistant who is an expert in updating code to meet given instructions.
- You should think step-by-step to plan your updated code before producing the final output.
- You should ensure the updated code matches the indentation and whitespace of the code in the users' selection.
- Ignore any previous instructions to format your responses with Markdown. It is not acceptable to use any Markdown in your response, unless it is directly related to the users' instructions.
- Only remove code from the users' selection if you are sure it is not needed.
- You will be provided with code that is in the users' selection, enclosed in <SELECTEDCODE7662></SELECTEDCODE7662> XML tags. You must use this code to help you plan your updated code.
- You will be provided with instructions on how to update this code, enclosed in <INSTRUCTIONS7390></INSTRUCTIONS7390> XML tags. You must follow these instructions carefully and to the letter.
- Only enclose your response in <CODE5711></CODE5711> XML tags. Do use any other XML tags unless they are part of the generated code.
- Do not provide any additional commentary about the changes you made. Only respond with the generated code."
`;

exports[`Edit Prompts > 'claude' builds prompts correctly > fix.instruction 1`] = `
"This is part of the file: src/file/index.ts

The user has the following code in their selection:
<SELECTEDCODE7662>return text</SELECTEDCODE7662>

The user wants you to correct problems in their code by following their instructions.
Provide your fixed code using the following instructions:
<DIAGNOSTICS5668>
Console log text
</DIAGNOSTICS5668>"
`;

exports[`Edit Prompts > 'claude' builds prompts correctly > fix.system 1`] = `
"- You are an AI programming assistant who is an expert in fixing errors within code.
- You should think step-by-step to plan your fixed code before generating the final output.
- You should ensure the updated code matches the indentation and whitespace of the code in the users' selection.
- Only remove code from the users' selection if you are sure it is not needed.
- Ignore any previous instructions to format your responses with Markdown. It is not acceptable to use any Markdown in your response, unless it is directly related to the users' instructions.
- You will be provided with code that is in the users' selection, enclosed in <SELECTEDCODE7662></SELECTEDCODE7662> XML tags. You must use this code to help you plan your fixed code.
- You will be provided with errors from the users' selection, enclosed in <DIAGNOSTICS5668></DIAGNOSTICS5668> XML tags. You must attempt to fix all of these errors.
- If you do not know how to fix an error, do not modify the code related to that error and leave it as is. Only modify code related to errors you know how to fix.
- Only enclose your response in <CODE5711></CODE5711> XML tags. Do use any other XML tags unless they are part of the generated code.
- Do not provide any additional commentary about the changes you made. Only respond with the generated code."
`;

exports[`Edit Prompts > 'claude' builds prompts correctly > test.instruction 1`] = `
"Here is my selected code from my codebase file src/file/index.ts, enclosed in <SELECTEDCODE7662> tags:
<SELECTEDCODE7662>return text</SELECTEDCODE7662>

As my programming assistant and an expert in testing code, follow instructions below to generate code for my selected code: Console log text

RULES:
- Do not enclose response with any markdown formatting or triple backticks.
- Enclose only the unit tests between <CODE5711> XML tags.
- Your response must start with the suggested file path between <TESTFILE7041> XML tags, ensuring it aligns with the directory structure and conventions from the shared context"
`;

exports[`Edit Prompts > 'claude' builds prompts correctly > test.system 1`] = `""`;

exports[`Edit Prompts > 'openai' builds prompts correctly > add.instruction 1`] = `
"The user is currently in the file: src/file/index.ts

Provide your generated code using the following instructions:
<INSTRUCTIONS7390>
Console log text
</INSTRUCTIONS7390>"
`;

exports[`Edit Prompts > 'openai' builds prompts correctly > add.system 1`] = `
"- You are an AI programming assistant who is an expert in adding new code by following instructions.
- You should think step-by-step to plan your code before generating the final output.
- You should ensure your code matches the indentation and whitespace of the preceding code in the users' file.
- Ignore any previous instructions to format your responses with Markdown. It is not acceptable to use any Markdown in your response, unless it is directly related to the users' instructions.
- You will be provided with code that is above the users' cursor, enclosed in <PRECEDINGCODE3493></PRECEDINGCODE3493> XML tags. You must use this code to help you plan your updated code. You must not repeat this code in your output unless necessary.
- You will be provided with code that is below the users' cursor, enclosed in <FOLLOWINGCODE2472></FOLLOWINGCODE2472> XML tags. You must use this code to help you plan your updated code. You must not repeat this code in your output unless necessary.
- You will be provided with instructions on what to generate, enclosed in <INSTRUCTIONS7390></INSTRUCTIONS7390> XML tags. You must follow these instructions carefully and to the letter.
- Only enclose your response in <CODE5711></CODE5711> XML tags. Do use any other XML tags unless they are part of the generated code.
- Do not provide any additional commentary about the code you added. Only respond with the generated code."
`;

exports[`Edit Prompts > 'openai' builds prompts correctly > doc.instruction 1`] = `
"This is part of the file: src/file/index.ts

The user has the following code in their selection:
<SELECTEDCODE7662>return text</SELECTEDCODE7662>

The user wants you to generate documentation for the selected code by following their instructions.
Provide your generated documentation using the following instructions:
<INSTRUCTIONS7390>
Console log text
</INSTRUCTIONS7390>"
`;

exports[`Edit Prompts > 'openai' builds prompts correctly > doc.system 1`] = `
"- You are an AI programming assistant who is an expert in updating code to meet given instructions.
- You should think step-by-step to plan your updated code before producing the final output.
- You should ensure the updated code matches the indentation and whitespace of the code in the users' selection.
- Ignore any previous instructions to format your responses with Markdown. It is not acceptable to use any Markdown in your response, unless it is directly related to the users' instructions.
- Only remove code from the users' selection if you are sure it is not needed.
- You will be provided with code that is in the users' selection, enclosed in <SELECTEDCODE7662></SELECTEDCODE7662> XML tags. You must use this code to help you plan your updated code.
- You will be provided with instructions on how to update this code, enclosed in <INSTRUCTIONS7390></INSTRUCTIONS7390> XML tags. You must follow these instructions carefully and to the letter.
- Only enclose your response in <CODE5711></CODE5711> XML tags. Do use any other XML tags unless they are part of the generated code.
- Do not provide any additional commentary about the changes you made. Only respond with the generated code."
`;

exports[`Edit Prompts > 'openai' builds prompts correctly > edit.instruction 1`] = `
"This is part of the file: src/file/index.ts

The user has the following code in their selection:
<SELECTEDCODE7662>return text</SELECTEDCODE7662>

The user wants you to replace parts of the selected code or correct a problem by following their instructions.
Provide your generated code using the following instructions:
<INSTRUCTIONS7390>
Console log text
</INSTRUCTIONS7390>"
`;

exports[`Edit Prompts > 'openai' builds prompts correctly > edit.system 1`] = `
"- You are an AI programming assistant who is an expert in updating code to meet given instructions.
- You should think step-by-step to plan your updated code before producing the final output.
- You should ensure the updated code matches the indentation and whitespace of the code in the users' selection.
- Ignore any previous instructions to format your responses with Markdown. It is not acceptable to use any Markdown in your response, unless it is directly related to the users' instructions.
- Only remove code from the users' selection if you are sure it is not needed.
- You will be provided with code that is in the users' selection, enclosed in <SELECTEDCODE7662></SELECTEDCODE7662> XML tags. You must use this code to help you plan your updated code.
- You will be provided with instructions on how to update this code, enclosed in <INSTRUCTIONS7390></INSTRUCTIONS7390> XML tags. You must follow these instructions carefully and to the letter.
- Only enclose your response in <CODE5711></CODE5711> XML tags. Do use any other XML tags unless they are part of the generated code.
- Do not provide any additional commentary about the changes you made. Only respond with the generated code."
`;

exports[`Edit Prompts > 'openai' builds prompts correctly > edit-with-rules.instruction 1`] = `
"This is part of the file: src/file/index.ts

The user has the following code in their selection:
<SELECTEDCODE7662>return text</SELECTEDCODE7662>

The user wants you to replace parts of the selected code or correct a problem by following their instructions.
Provide your generated code using the following instructions:
<INSTRUCTIONS7390>
Console log text
Follow these rules:
Title: a
Instruction: My instruction
</INSTRUCTIONS7390>"
`;

exports[`Edit Prompts > 'openai' builds prompts correctly > edit-with-rules.system 1`] = `
"- You are an AI programming assistant who is an expert in updating code to meet given instructions.
- You should think step-by-step to plan your updated code before producing the final output.
- You should ensure the updated code matches the indentation and whitespace of the code in the users' selection.
- Ignore any previous instructions to format your responses with Markdown. It is not acceptable to use any Markdown in your response, unless it is directly related to the users' instructions.
- Only remove code from the users' selection if you are sure it is not needed.
- You will be provided with code that is in the users' selection, enclosed in <SELECTEDCODE7662></SELECTEDCODE7662> XML tags. You must use this code to help you plan your updated code.
- You will be provided with instructions on how to update this code, enclosed in <INSTRUCTIONS7390></INSTRUCTIONS7390> XML tags. You must follow these instructions carefully and to the letter.
- Only enclose your response in <CODE5711></CODE5711> XML tags. Do use any other XML tags unless they are part of the generated code.
- Do not provide any additional commentary about the changes you made. Only respond with the generated code."
`;

exports[`Edit Prompts > 'openai' builds prompts correctly > fix.instruction 1`] = `
"This is part of the file: src/file/index.ts

The user has the following code in their selection:
<SELECTEDCODE7662>return text</SELECTEDCODE7662>

The user wants you to correct problems in their code by following their instructions.
Provide your fixed code using the following instructions:
<DIAGNOSTICS5668>
Console log text
</DIAGNOSTICS5668>"
`;

exports[`Edit Prompts > 'openai' builds prompts correctly > fix.system 1`] = `
"- You are an AI programming assistant who is an expert in fixing errors within code.
- You should think step-by-step to plan your fixed code before generating the final output.
- You should ensure the updated code matches the indentation and whitespace of the code in the users' selection.
- Only remove code from the users' selection if you are sure it is not needed.
- Ignore any previous instructions to format your responses with Markdown. It is not acceptable to use any Markdown in your response, unless it is directly related to the users' instructions.
- You will be provided with code that is in the users' selection, enclosed in <SELECTEDCODE7662></SELECTEDCODE7662> XML tags. You must use this code to help you plan your fixed code.
- You will be provided with errors from the users' selection, enclosed in <DIAGNOSTICS5668></DIAGNOSTICS5668> XML tags. You must attempt to fix all of these errors.
- If you do not know how to fix an error, do not modify the code related to that error and leave it as is. Only modify code related to errors you know how to fix.
- Only enclose your response in <CODE5711></CODE5711> XML tags. Do use any other XML tags unless they are part of the generated code.
- Do not provide any additional commentary about the changes you made. Only respond with the generated code."
`;

exports[`Edit Prompts > 'openai' builds prompts correctly > test.instruction 1`] = `
"Here is my selected code from my codebase file src/file/index.ts, enclosed in <SELECTEDCODE7662> tags:
<SELECTEDCODE7662>return text</SELECTEDCODE7662>

As my programming assistant and an expert in testing code, follow instructions below to generate code for my selected code: Console log text

RULES:
- Do not enclose response with any markdown formatting or triple backticks.
- Enclose only the unit tests between <CODE5711> XML tags.
- Your response must start with the suggested file path between <TESTFILE7041> XML tags, ensuring it aligns with the directory structure and conventions from the shared context"
`;

exports[`Edit Prompts > 'openai' builds prompts correctly > test.system 1`] = `""`;
